File: chess/piece.py
from enum import Enum
from pygame.rect import Rect
from pygame.surface import Surface


class Color(Enum):
    WHITE = 0
    BLACK = 1

    def texture_coordinate(self) -> float:
        return self.value / 2


class Kind(Enum):
    KING = 0
    QUEEN = 1
    BISHOP = 2
    KNIGHT = 3
    ROOK = 4
    PAWN = 5

    def texture_coordinate(self) -> float:
        return self.value / 6


class Piece:
    def __init__(self, color: Color, kind: Kind) -> None:
        self.color = color
        self.kind = kind

    def __str__(self) -> str:
        return ""

    def draw(self, surface: Surface, t: Surface) -> None:
        x_ratio = self.kind.texture_coordinate()
        y_ratio = self.color.texture_coordinate()

        rect = (t.get_width() * x_ratio, t.get_height() * y_ratio, t.get_width() / 6, t.get_height() / 2)

        surface.blit(t, (0, 0), rect)

File: src/piece.rs
pub enum PieceColor {
    Black,
    White,
}

pub enum PieceKind {
    Bishop,
    King,
    Knight,
    Pawn,
    Queen,
    Rook,
}

pub struct Piece {
    color: PieceColor,
    kind: PieceKind,
}

impl Piece {
    pub fn new(color: PieceColor, kind: PieceKind) -> Self {
        Self { color, kind }
    }

    pub fn from_char(c: char) -> Option<Self> {
        use PieceColor as Color;
        use PieceKind as Kind;

        match c {
            'b' => Some(Piece::new(Color::Black, Kind::Bishop)),
            'k' => Some(Piece::new(Color::Black, Kind::King)),
            'n' => Some(Piece::new(Color::Black, Kind::Knight)),
            'p' => Some(Piece::new(Color::Black, Kind::Pawn)),
            'q' => Some(Piece::new(Color::Black, Kind::Queen)),
            'r' => Some(Piece::new(Color::Black, Kind::Rook)),
            'B' => Some(Piece::new(Color::White, Kind::Bishop)),
            'K' => Some(Piece::new(Color::White, Kind::King)),
            'N' => Some(Piece::new(Color::White, Kind::Knight)),
            'P' => Some(Piece::new(Color::White, Kind::Pawn)),
            'Q' => Some(Piece::new(Color::White, Kind::Queen)),
            'R' => Some(Piece::new(Color::White, Kind::Rook)),
            _ => None,
        }
    }
}

impl std::str::FromStr for Piece {
    type Err = ();

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        use PieceColor as PC;
        use PieceKind as PK;

        match s {
            "b" => Ok(Piece::new(PC::Black, PK::Bishop)),
            "k" => Ok(Piece::new(PC::Black, PK::King)),
            "n" => Ok(Piece::new(PC::Black, PK::Knight)),
            "p" => Ok(Piece::new(PC::Black, PK::Pawn)),
            "q" => Ok(Piece::new(PC::Black, PK::Queen)),
            "r" => Ok(Piece::new(PC::Black, PK::Rook)),
            "B" => Ok(Piece::new(PC::White, PK::Bishop)),
            "K" => Ok(Piece::new(PC::White, PK::King)),
            "N" => Ok(Piece::new(PC::White, PK::Knight)),
            "P" => Ok(Piece::new(PC::White, PK::Pawn)),
            "Q" => Ok(Piece::new(PC::White, PK::Queen)),
            "R" => Ok(Piece::new(PC::White, PK::Rook)),
            _ => Err(()),
        }
    }
}
